File: ios/Classes/proxy/MediaStreamTrackProxy.swift
import WebRTC

/// Wrapper around a `MediaStreamTrack`, powering it with additional API.
class MediaStreamTrackProxy: Equatable {
  /// Indicator whether the `stop()` was called on this `MediaStreamTrackProxy`.
  private var isStopped = false

  /// Device ID of this `MediaStreamTrackProxy`.
  private var deviceId = "remote"

  /// Source of this `MediaStreamTrackProxy`.
  private var source: MediaTrackSource?

  /// Underlying `RTCMediaStreamTrack`.
  private var track: RTCMediaStreamTrack

  /// Subscribers for `onStop` callback of this `MediaStreamTrackProxy`.
  private var onStopSubscribers: [() -> Void] = []

  /// Subscribers for `onEnded` callback of this `MediaStreamTrackProxy`.
  private var onEndedSubscribers: [() -> Void] = []

  /// `RTCVideoRenderer` that tracks video track height and width changes.
  private var dimensionsSink: VideoTrackDimensionsSink?

  /// `RTCVideoRenderer` that tracks video track height and width changes.
  class VideoTrackDimensionsSink: NSObject, RTCVideoRenderer {
    /// Lock that protects width and height.
    let cond = NSCondition()

    /// Latest video frame width.
    var width: Int?

    /// Latest video frame height.
    var height: Int?

    /// Blocks current thread until width and height values are set, should
    /// be called before accessing these values to ensure correct
    /// synchronization.
    func waitForVideoDimensions() {
      self.cond.lock()
      while self.width == nil || self.height == nil {
        self.cond.wait()
      }
      self.cond.unlock()
    }

    /// Fetches width and height values
    func renderFrame(_ renderFrame: RTCVideoFrame?) {
      let width = Int(renderFrame!.buffer.width)
      let height = Int(renderFrame!.buffer.height)

      if height == 0 || width == 0 {
        return
      }

      self.cond.lock()

      self.width = width
      self.height = height

      self.cond.broadcast()
      self.cond.unlock()
    }

    func setSize(_: CGSize) {}
  }

  /// Initializes a new `MediaStreamTrackProxy` based on the provided data.
  init(
    track: RTCMediaStreamTrack,
    deviceId: String?,
    source: MediaTrackSource?
  ) {
    self.source = source
    if deviceId != nil {
      self.deviceId = deviceId!
    }
    self.track = track
    MediaStreamTrackStore.tracks[track.trackId] = self

    if track.kind == "video" {
      self.dimensionsSink = VideoTrackDimensionsSink()
      let videoTrack = self.track as! RTCVideoTrack
      videoTrack.add(self.dimensionsSink!)
    }
  }

  /// Compares two `MediaStreamTrackProxy`s based on underlying
  /// `RTCMediaStreamTrack`s.
  static func == (lhs: MediaStreamTrackProxy,
                  rhs: MediaStreamTrackProxy) -> Bool
  {
    lhs.track == rhs.track
  }

  /// Adds the specified `RTCVideoRenderer` to the underlying
  /// `RTCMediaStreamTrack`.
  func addRenderer(renderer: RTCVideoRenderer) {
    let videoTrack = self.track as! RTCVideoTrack
    videoTrack.add(renderer)
  }

  /// Removes the specified `RTCVideoRenderer` from the underlying
  /// `RTCMediaStreamTrack`.
  func removeRenderer(renderer: RTCVideoRenderer) {
    let videoTrack = self.track as! RTCVideoTrack
    videoTrack.remove(renderer)
  }

  /// Returns ID of this track.
  func id() -> String {
    self.track.trackId
  }

  /// Returns `MediaType` of this track.
  func kind() -> MediaType {
    let kind = self.track.kind
    switch kind {
    case "audio":
      return MediaType.audio
    case "video":
      return MediaType.video
    default:
      abort()
    }
  }

  /// Returns latest frame width if this is a video track or nil otherwise.
  func getWidth() async -> Int? {
    if self.dimensionsSink == nil {
      return nil
    } else {
      let task = Task.detached {
        self.dimensionsSink!.waitForVideoDimensions()
        return self.dimensionsSink!.width!
      }
      return await task.get()
    }
  }

  /// Returns latest frame height if this is a video track or nil otherwise.
  func getHeight() async -> Int? {
    if self.dimensionsSink == nil {
      return nil
    } else {
      let task = Task.detached {
        self.dimensionsSink!.waitForVideoDimensions()
        return self.dimensionsSink!.height!
      }
      return await task.get()
    }
  }

  /// Returns `FacingMode` of this `MediaStreamTrackProxy`.
  ///
  /// Returns `nil` if it's an audio track.
  func getFacingMode() -> FacingMode? {
    let source = self.source as? VideoMediaTrackSourceProxy
    if source != nil {
      return source!.getFacingMode()
    } else {
      return nil
    }
  }

  /// Returns device ID of this track.
  func getDeviceId() -> String {
    self.deviceId
  }

  /// Returns a forked `MediaStreamTrackProxy` based on the same source as this
  /// track has.
  func fork() throws -> MediaStreamTrackProxy {
    if self.source == nil {
      throw MediaStreamTrackException.remoteTrackCantBeCloned
    } else {
      return self.source!.newTrack()
    }
  }

  /// Stops this track.
  ///
  /// Source will be stopped and disposed once all its tracks are stopped.
  func stop() {
    self.isStopped = true
    for cb in self.onStopSubscribers {
      cb()
    }
  }

  /// Returns the current `readyState` of this track.
  func state() -> MediaStreamTrackState {
    let state = self.track.readyState
    switch state {
    case .live:
      return MediaStreamTrackState.live
    case .ended:
      return MediaStreamTrackState.ended
    default:
      abort()
    }
  }

  /// Sets `enabled` state of this track.
  func setEnabled(enabled: Bool) {
    self.track.isEnabled = enabled
  }

  /// Subscribes to `onStopped` callback of this track.
  func onStopped(cb: @escaping () -> Void) {
    self.onStopSubscribers.append(cb)
  }

  /// Notifies `RtpReceiverProxy` about its `MediaStreamTrackProxy` being
  /// removed from the receiver.
  func notifyEnded() {
    if self.track.readyState == .ended {
      for cb in self.onEndedSubscribers {
        cb()
      }
    }
  }

  /// Subscribes to `onEnded` callback of this track.
  func onEnded(cb: @escaping () -> Void) {
    self.onEndedSubscribers.append(cb)
  }

  /// Returns the underlying `RTCMediaStreamTrack` of this proxy.
  func obj() -> RTCMediaStreamTrack {
    self.track
  }
}

File: android/src/main/kotlin/com/instrumentisto/medea_flutter_webrtc/proxy/MediaStreamTrackProxy.kt
package com.instrumentisto.medea_flutter_webrtc.proxy

import android.util.Log
import com.instrumentisto.medea_flutter_webrtc.TrackRepository
import com.instrumentisto.medea_flutter_webrtc.model.FacingMode
import com.instrumentisto.medea_flutter_webrtc.model.MediaStreamTrackState
import com.instrumentisto.medea_flutter_webrtc.model.MediaType
import kotlinx.coroutines.CompletableDeferred
import org.webrtc.MediaStreamTrack
import org.webrtc.VideoSink
import org.webrtc.VideoTrack

private val TAG = MediaStreamTrackProxy::class.java.simpleName

/**
 * Wrapper around a [MediaStreamTrack].
 *
 * @param track Underlying [MediaStreamTrack].
 * @property deviceId Unique ID of device on which this [MediaStreamTrackProxy] is based.
 * @property source [MediaTrackSource] from which this [MediaStreamTrackProxy] is created. `null`
 * for [MediaStreamTrackProxy]s received from the remote side.
 */
class MediaStreamTrackProxy(
    track: MediaStreamTrack,
    val facingMode: FacingMode? = null,
    val deviceId: String = "remote",
    private val source: MediaTrackSource? = null
) : Proxy<MediaStreamTrack>(track) {
  /**
   * Subscribers for the [onStop] callback.
   *
   * Will be called once on a [stop] call.
   */
  private var onStopSubscribers: MutableList<() -> Unit> = mutableListOf()

  /** Indicates that this [stop] was called on this [MediaStreamTrackProxy]. */
  private var isStopped: Boolean = false

  /** List of [EventObserver]s belonging to this [MediaStreamTrackProxy]. */
  private var eventObservers: HashSet<EventObserver> = HashSet()

  /** Indicator whether the underlying [MediaStreamTrack] had been disposed. */
  private var disposed: Boolean = false

  /** [VideoSink] that tracks height and width changes. */
  private var sink: VideoSink? = null

  /** Provides asynchronous wait for [width] and [height] initialization. */
  private val fetchDimensions = CompletableDeferred<Unit>()

  /** Video width */
  @Volatile private var width: Int = 0

  /** Video height */
  @Volatile private var height: Int = 0

  /** [MediaType] of the underlying [MediaStreamTrack]. */
  val kind: MediaType =
      when (obj.kind()) {
        MediaStreamTrack.VIDEO_TRACK_KIND -> MediaType.VIDEO
        MediaStreamTrack.AUDIO_TRACK_KIND -> MediaType.AUDIO
        else -> throw Exception("LibWebRTC provided unknown MediaType value")
      }

  /** ID of the underlying [MediaStreamTrack] */
  val id: String = obj.id()

  /** [MediaStreamTrackState] of the underlying [MediaStreamTrack]. */
  var state: MediaStreamTrackState = MediaStreamTrackState.LIVE
    get() {
      field =
          if (disposed) {
            MediaStreamTrackState.ENDED
          } else {
            MediaStreamTrackState.fromWebRtcState(obj.state())
          }

      return field
    }

  init {
    TrackRepository.addTrack(this)

    if (kind == MediaType.VIDEO) {
      sink = VideoSink { frame ->
        width = frame.buffer.width
        height = frame.buffer.height
        fetchDimensions.complete(Unit)
      }
      (obj as VideoTrack).addSink(sink)

      addOnSyncListener { (obj as VideoTrack).addSink(sink) }
    }
  }

  /** Returns the video [width] of the track. */
  suspend fun width(): Int? {
    if (kind == MediaType.AUDIO) {
      return null
    }

    fetchDimensions.await()
    return width
  }

  /** Returns the video [height] of the track. */
  suspend fun height(): Int? {
    if (kind == MediaType.AUDIO) {
      return null
    }
    fetchDimensions.await()
    return height
  }

  /** Sets the [disposed] property to `true`. */
  fun setDisposed() {
    disposed = true
  }

  companion object {
    /** Observer of [MediaStreamTrackProxy] events. */
    interface EventObserver {
      fun onEnded()
    }
  }

  /**
   * Creates a broadcaster to all the [eventObservers] of this [MediaStreamTrackProxy].
   *
   * @return [EventObserver] broadcasting calls to all the [eventObservers].
   */
  fun observableEventBroadcaster(): EventObserver {
    return object : EventObserver {
      override fun onEnded() {
        eventObservers.forEach { it.onEnded() }
        eventObservers.clear()
      }
    }
  }

  /**
   * Adds the specified [EventObserver] to this [MediaStreamTrackProxy].
   *
   * @param eventObserver [EventObserver] to be subscribed.
   */
  fun addEventObserver(eventObserver: EventObserver) {
    eventObservers.add(eventObserver)
  }

  /**
   * Removes the specified [EventObserver] from the list of [EventObserver]s.
   *
   * @param eventObserver [EventObserver] to be removed.
   */
  fun removeEventObserver(eventObserver: EventObserver) {
    eventObservers.remove(eventObserver)
  }

  /**
   * Creates a new [MediaStreamTrackProxy] based on the same [MediaTrackSource] as this
   * [MediaStreamTrackProxy].
   *
   * Can be called only on local [MediaStreamTrackProxy]s.
   *
   * @throws Exception If called on a remote [MediaStreamTrackProxy].
   *
   * @return Created [MediaStreamTrackProxy].
   */
  fun fork(): MediaStreamTrackProxy {
    if (this.source == null) {
      throw Exception("Remote MediaStreamTracks can't be cloned")
    } else {
      return source.newTrack()
    }
  }

  /**
   * Stops this [MediaStreamTrackProxy].
   *
   * Media source will be disposed only if there is no another [MediaStreamTrackProxy] depending on
   * a [MediaTrackSource].
   */
  fun stop() {
    if (!isStopped) {
      isStopped = true
      onStopSubscribers.forEach { sub -> sub() }
      onStopSubscribers.clear()
      if (sink != null) {
        (obj as VideoTrack).removeSink(sink)
        sink = null
      }
    } else {
      Log.w(TAG, "Double stop detected [deviceId: $deviceId]!")
    }
  }

  /**
   * Sets enabled state of the underlying [MediaStreamTrack]. Does nothing if the underlying
   * [MediaStreamTrack] has been [disposed].
   *
   * @param enabled State which will be set to the underlying [MediaStreamTrack].
   */
  fun setEnabled(enabled: Boolean) {
    if (disposed) {
      return
    }

    obj.setEnabled(enabled)
  }

  /**
   * Subscribes to the [stop] event of this [MediaStreamTrackProxy].
   *
   * This callback is guaranteed to be called only once.
   */
  fun onStop(f: () -> Unit) {
    onStopSubscribers.add(f)
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as MediaStreamTrackProxy

    if (id != other.id) return false

    return true
  }

  override fun hashCode(): Int {
    return id.hashCode()
  }
}
